File: moxy-compiler/src/main/java/com/omegar/mvp/compiler/presenterbinder/TargetClassInfoToPresenterBinderJavaFileProcessor.java
package com.omegar.mvp.compiler.presenterbinder;

import com.omegar.mvp.MvpPresenter;
import com.omegar.mvp.MvpProcessor;
import com.omegar.mvp.PresenterBinder;
import com.omegar.mvp.compiler.entity.TargetClassInfo;
import com.omegar.mvp.compiler.entity.TargetPresenterField;
import com.omegar.mvp.compiler.pipeline.JavaFileProcessor;
import com.omegar.mvp.compiler.Util;
import com.omegar.mvp.presenter.PresenterField;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeSpec;
import com.squareup.javapoet.WildcardTypeName;

import java.util.ArrayList;
import java.util.List;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;

/**
 * 18.12.2015
 * <p>
 * Generates PresenterBinder for class annotated with &#64;InjectPresenters
 * <p>
 * for Sample class with single injected presenter
 * <pre>
 * {@code
 *
 * &#64;InjectPresenters
 * public class Sample extends MvpActivity implements MyView
 * {
 *
 * &#64;InjectPresenter(type = PresenterType.LOCAL, tag = "SOME_TAG")
 * com.arellomobile.example.MyPresenter mMyPresenter;
 *
 * }
 *
 * }
 * </pre>
 * <p>
 * PresenterBinderClassGenerator generates PresenterBinder
 * <p>
 *
 * @author Yuri Shmakov
 * @author Alexander Blinov
 */
@SuppressWarnings("NewApi")
public final class TargetClassInfoToPresenterBinderJavaFileProcessor extends JavaFileProcessor<TargetClassInfo> {

	@Override
	public JavaFile process(TargetClassInfo targetClassInfo) {
		ClassName targetClassName = targetClassInfo.getName();
		TypeElement element = targetClassInfo.getTypeElement();
		List<TargetPresenterField> fields = targetClassInfo.getFields();

		final String containerSimpleName = String.join("$", targetClassName.simpleNames());

		TypeSpec.Builder classBuilder = TypeSpec.classBuilder(containerSimpleName + MvpProcessor.PRESENTER_BINDER_SUFFIX)
				.addOriginatingElement(targetClassInfo.getTypeElement())
				.addModifiers(Modifier.PUBLIC)
				.superclass(ParameterizedTypeName.get(ClassName.get(PresenterBinder.class), targetClassName));

		for (TargetPresenterField field : fields) {
			classBuilder.addType(generatePresenterBinderClass(element, field, targetClassName));
		}

		classBuilder.addMethod(generateGetPresentersMethod(fields, targetClassName));

		return JavaFile.builder(targetClassName.packageName(), classBuilder.build())
				.indent("\t")
				.build();
	}

	private static MethodSpec generateGetPresentersMethod(List<TargetPresenterField> fields,
	                                                      ClassName containerClassName) {
		MethodSpec.Builder builder = MethodSpec.methodBuilder("getPresenterFields")
				.addAnnotation(Override.class)
				.addModifiers(Modifier.PUBLIC)
				.returns(ParameterizedTypeName.get(
						ClassName.get(List.class), ParameterizedTypeName.get(
								ClassName.get(PresenterField.class), containerClassName)));

		builder.addStatement("$T<$T<$T>> presenters = new $T<>($L)",
				List.class, PresenterField.class, containerClassName,
				ArrayList.class, fields.size());

		for (TargetPresenterField field : fields) {
			builder.addStatement("presenters.add(new $L())", field.getGeneratedClassName());
		}

		builder.addStatement("return presenters");

		return builder.build();
	}

	private static TypeSpec generatePresenterBinderClass(TypeElement element, TargetPresenterField field,
														 ClassName targetClassName) {
		String tag = field.getTag();
		if (tag == null) tag = field.getName();

		TypeSpec.Builder classBuilder = TypeSpec.classBuilder(field.getGeneratedClassName())
				.addOriginatingElement(element)
				.addModifiers(Modifier.PUBLIC)
				.superclass(ParameterizedTypeName.get(
						ClassName.get(PresenterField.class), targetClassName))
				.addMethod(generatePresenterBinderConstructor(field, tag))
				.addMethod(generateBindMethod(field, targetClassName))
				.addMethod(generateProvidePresenterMethod(field, targetClassName));

		String tagProviderMethodName = field.getPresenterTagProviderMethodName();
		if (tagProviderMethodName != null) {
			classBuilder.addMethod(generateGetTagMethod(tagProviderMethodName, targetClassName));
		}

		return classBuilder.build();
	}

	private static MethodSpec generatePresenterBinderConstructor(TargetPresenterField field, String tag) {
		return MethodSpec.constructorBuilder()
				.addModifiers(Modifier.PUBLIC)
				.addStatement("super($S, $T.$L, $S, $T.class)",
						tag,
						field.getPresenterType().getDeclaringClass(),
						field.getPresenterType().name(),
						field.getPresenterId(),
						field.getTypeName())
				.build();
	}

	private static MethodSpec generateBindMethod(TargetPresenterField field,
												 ClassName targetClassName) {
		return MethodSpec.methodBuilder("bind")
				.addAnnotation(Override.class)
				.addModifiers(Modifier.PUBLIC)
				.addParameter(targetClassName, "target")
				.addParameter(MvpPresenter.class, "presenter")
				.addStatement("target.$L = ($T) presenter", field.getName(), field.getTypeName())
				.build();
	}

	private static MethodSpec generateProvidePresenterMethod(TargetPresenterField field,
															 ClassName targetClassName) {
		MethodSpec.Builder builder = MethodSpec.methodBuilder("providePresenter")
				.addAnnotation(Override.class)
				.addModifiers(Modifier.PUBLIC)
				.returns(ParameterizedTypeName.get(
						ClassName.get(MvpPresenter.class), WildcardTypeName.subtypeOf(Object.class)))
				.addParameter(targetClassName, "delegated");

		if (field.getPresenterProviderMethodName() != null) {
			builder.addStatement("return delegated.$L()", field.getPresenterProviderMethodName());
		} else {
			boolean hasEmptyConstructor = Util.hasEmptyConstructor((TypeElement) ((DeclaredType) field.getClazz()).asElement());

			if (hasEmptyConstructor) {
				builder.addStatement("return new $T()", field.getTypeName());
			} else {
				builder.addStatement(
						"throw new $T($S + $S)", IllegalStateException.class,
						field.getClazz(), " has not default constructor. You can apply @ProvidePresenter to some method which will construct Presenter. Also you can make it default constructor");
			}
		}

		return builder.build();
	}

	private static MethodSpec generateGetTagMethod(String tagProviderMethodName,
	                                               ClassName targetClassName) {
		return MethodSpec.methodBuilder("getTag")
				.addAnnotation(Override.class)
				.addModifiers(Modifier.PUBLIC)
				.returns(String.class)
				.addParameter(targetClassName, "delegated")
				.addStatement("return String.valueOf(delegated.$L())", tagProviderMethodName)
				.build();
	}
}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/presenterbinder/TargetClassInfoToPresenterBinderJavaFileProcessor.kt
package com.omegar.mvp.compiler.presenterbinder

import com.omegar.mvp.MvpPresenter
import com.omegar.mvp.PresenterBinder
import com.omegar.mvp.compiler.MoxyConst
import com.omegar.mvp.compiler.Util
import com.omegar.mvp.compiler.entity.TargetClassInfo
import com.omegar.mvp.compiler.entity.TargetPresenterField
import com.omegar.mvp.compiler.pipeline.KotlinFile
import com.omegar.mvp.compiler.pipeline.KotlinFileProcessor
import com.omegar.mvp.presenter.PresenterField
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import javax.lang.model.element.TypeElement
import javax.lang.model.type.DeclaredType

/**
 * 18.12.2015
 *
 *
 * Generates PresenterBinder for class annotated with &#64;InjectPresenters
 *
 *
 * for Sample class with single injected presenter
 * <pre>
 * `&#64;InjectPresenters
 * public class Sample extends MvpActivity implements MyView
 * {
 *
 * &#64;InjectPresenter(type = PresenterType.LOCAL, tag = "SOME_TAG")
 * com.arellomobile.example.MyPresenter mMyPresenter;
 *
 * }
 *
` *
</pre> *
 *
 *
 * PresenterBinderClassGenerator generates PresenterBinder
 *
 *
 *
 * @author Yuri Shmakov
 * @author Alexander Blinov
 */
class TargetClassInfoToPresenterBinderJavaFileProcessor : KotlinFileProcessor<TargetClassInfo>() {

    override fun process(targetClassInfo: TargetClassInfo): KotlinFile {
        val targetClassName = targetClassInfo.name
        val element: TypeElement = targetClassInfo.typeElement
        val fields = targetClassInfo.fields

        val className = targetClassName.simpleNames.joinToString("$") + MoxyConst.PRESENTER_BINDER_SUFFIX
        val classBuilder = TypeSpec.classBuilder(className)
            .addOriginatingElement(targetClassInfo.typeElement)
            .superclass(PresenterBinder::class.asClassName().parameterizedBy(targetClassName, MvpPresenter::class.asClassName().parameterizedBy(STAR)))
            .addProperty(generateGetPresentersMethod(fields, targetClassName))
            .addTypes(fields.map { field ->
                generatePresenterBinderClass(element, field, targetClassName)
            })


        return FileSpec.builder(targetClassName.packageName, className)
            .addType(classBuilder.build())
            .indent("\t")
            .build()
    }

    private fun generateGetPresentersMethod(
        fields: List<TargetPresenterField>,
        containerClassName: ClassName
    ): PropertySpec {
        val listTypeName = LIST.parameterizedBy(
            PresenterField::class.asClassName().parameterizedBy(
                containerClassName,
                MvpPresenter::class.asClassName().parameterizedBy(STAR)
            )
        )
        return PropertySpec.builder(
            "presenterFields",
            listTypeName,
            KModifier.OVERRIDE
        )
            .initializer(
                format = "listOf(" + fields.joinToString { "%L()" } + ") as %T",
                args = (fields.map { ClassName.bestGuess(it.generatedClassName) } + listTypeName).toTypedArray()
            )
            .build()
    }

    private fun generatePresenterBinderClass(
        element: TypeElement,
        field: TargetPresenterField,
        targetClassName: ClassName
    ): TypeSpec {
        val presenterTypeName = field.typeName
        val classBuilder = TypeSpec.classBuilder(field.generatedClassName)
            .addOriginatingElement(element)
            .addModifiers(KModifier.PRIVATE)
            .superclass(PresenterField::class.asClassName().parameterizedBy(targetClassName, presenterTypeName))
            .addSuperclassConstructorParameter("%T.%L", field.presenterType.declaringClass, field.presenterType.name)
            .addSuperclassConstructorParameter("%T::class", presenterTypeName)
            .addFunction(generateBindMethod(field, targetClassName, presenterTypeName))
            .addFunction(generateProvidePresenterMethod(field, targetClassName, presenterTypeName))
        val tagProviderMethodName = field.presenterTagProviderMethodName
        if (tagProviderMethodName != null) {
            classBuilder.addFunction(generateGetTagMethod(tagProviderMethodName, targetClassName))
        }
        return classBuilder.build()
    }

    private fun generateBindMethod(
        field: TargetPresenterField,
        targetClassName: ClassName,
        presenterTypeName: TypeName
    ): FunSpec {
        return FunSpec.builder("bind")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("target", targetClassName)
            .addParameter("presenter", presenterTypeName)
            .addStatement("target.%L = presenter", field.name)
            .build()
    }

    private fun generateProvidePresenterMethod(
        field: TargetPresenterField,
        targetClassName: ClassName,
        presenterTypeName: TypeName
    ): FunSpec {
        val builder = FunSpec.builder("providePresenter")
            .addModifiers(KModifier.OVERRIDE)
            .returns(presenterTypeName)
            .addParameter("delegated", targetClassName)
        if (field.presenterProviderMethodName != null) {
            builder.addStatement("return delegated.%L()", field.presenterProviderMethodName!!)
        } else {
            val hasEmptyConstructor = Util.hasEmptyConstructor((field.clazz as DeclaredType).asElement() as TypeElement)
            if (hasEmptyConstructor) {
                builder.addStatement("return %T()", field.typeName)
            } else {
                builder.addStatement(
                    "throw %T(%S + %S)",
                    IllegalStateException::class.java,
                    field.clazz,
                    " has not default constructor. You can apply @ProvidePresenter to some method which will construct Presenter. Also you can make it default constructor"
                )
            }
        }
        return builder.build()
    }

    private fun generateGetTagMethod(
        tagProviderMethodName: String,
        targetClassName: ClassName
    ): FunSpec {
        return FunSpec.builder("getTag")
            .addModifiers(KModifier.OVERRIDE)
            .returns(STRING)
            .addParameter("delegated", targetClassName)
            .addStatement("return delegated.%L.toString()", tagProviderMethodName)
            .build()
    }

}